### PR TITLE
fix(rollapp): fix register denom metadata not triggering VFC creation

### LINKED ISSUE #726: rollapp denometada registration doesn't tigger erc20 vfc creation
this is due to the fact it's being set not by using the denommetadata keeper by directly through the bank. this should be changed.

### PR DESCRIPTION
**Pull request by zale144**
## Description

Currently the new denoms don't get registered in the ibc transfer, going through the ibc_middleware. Instead, the middleware is removed and the registration will rely on the Genesis Event flow.

----

Closes #726 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/rollapp/keeper/keeper.go [hunk 1]
```diff
@@ -70,67 +74,40 @@ func (k Keeper) TriggerRollappGenesisEvent(ctx sdk.Context, rollapp types.Rollap
...
-			continue
-		}
 		// create a new token denom metadata where it's base = ibcDenom,
 		// and the rest of the fields are taken from rollapp.metadata
 		metadata := banktypes.Metadata{
 			Description: "auto-generated metadata for " + ibcBaseDenom + " from rollapp " + rollapp.RollappId,
 			Base:        ibcBaseDenom,
 			DenomUnits:  make([]*banktypes.DenomUnit, len(rollapp.TokenMetadata[i].DenomUnits)),
-			Display:     rollapp.TokenMetadata[i].Display,
+			Display:     ibcBaseDenom,
```
Comment (zale144) [inline x/rollapp/keeper/keeper.go:110]: @omritoptix Not too sure about this, but this is expected further down the line - denom display should match the 0th unit display

File: x/rollapp/keeper/keeper.go [hunk 2]
```diff
@@ -22,10 +23,11 @@ type (
...
 		paramstore paramtypes.Subspace
 
-		ibcclientKeeper types.IBCClientKeeper
-		transferKeeper  types.TransferKeeper
-		channelKeeper   types.ChannelKeeper
-		bankKeeper      types.BankKeeper
+		ibcclientKeeper     types.IBCClientKeeper
+		transferKeeper      types.TransferKeeper
+		channelKeeper       types.ChannelKeeper
+		bankKeeper          types.BankKeeper
```
[the remark this answers is not in the document]
Comment (mtsitrin) [reply x/rollapp/keeper/keeper.go:29]: and add the expected interfaces for `denommetadatamodulekeeper`

File: x/rollapp/keeper/keeper.go (hunk 1 above)
Comment (mtsitrin) [reply x/rollapp/keeper/keeper.go:110]: no, why is that? quite the opposite

Comment (zale144) [reply x/rollapp/keeper/keeper.go:110]: without this change, the error I get is `metadata must contain a denomination unit with display denom 'arax'`

**Review by mtsitrin — DISMISSED** [review]

File: ibctesting/denom_metadata_test.go [hunk 3]
```diff
@@ -90,9 +99,4 @@ func (suite *DenomMetaDataTestSuite) TestDenomRegistationRollappToHub() {
 	// relay send
 	err = path.RelayPacket(packet)
 	suite.Require().NoError(err)
-
-	metadata, found := ConvertToApp(suite.hubChain).BankKeeper.GetDenomMetaData(suite.hubChain.GetContext(), stakeVoucherDenom.IBCDenom())
-	suite.Require().True(found)
-	suite.Equal("bigstake", metadata.Display)
-	suite.Equal("BIGSTAKE", metadata.Symbol)
 }
```
Comment (danwt) [inline ibctesting/denom_metadata_test.go:102]: if these lines are removed, then what is the purpose of lines 89-102?

File: ibctesting/rollapp_genesis_token_test.go [hunk 4]
```diff
@@ -223,11 +222,8 @@ func (suite *RollappGenesisTokenTestSuite) TestTriggerGenesisEvent() {
 					}
 				}
 
-				denomMetaData, found := bankKeeper.GetDenomMetaData(suite.ctx, rollappIBCDenom)
-				suite.Require().Equal(tc.expectSavedDenomMetaData, found)
-				if tc.expectSavedDenomMetaData {
-					suite.Require().Equal(rollappDenom, denomMetaData.Display)
-				}
+				_, found := bankKeeper.GetDenomMetaData(suite.ctx, rollappIBCDenom)
+				suite.Require().Equal(tc.expectSavedDenom != "", found)
```
Comment (danwt) [inline ibctesting/rollapp_genesis_token_test.go:226]: does this need to be in the for loop?

File: x/rollapp/keeper/keeper.go [hunk 5]
```diff
@@ -70,60 +73,33 @@ func (k Keeper) TriggerRollappGenesisEvent(ctx sdk.Context, rollapp types.Rollap
...
+
+	// Register the denom metadata
+	if err := k.registerDenomMetadata(ctx, rollapp); err != nil {
+		return fmt.Errorf("failed to register denom metadata: %w", err)
+	}
+
 	// Call the mint genesis tokens function
 	if err := k.mintRollappGenesisTokens(ctx, rollapp); err != nil {
-		return err
+		return fmt.Errorf("failed to mint genesis tokens: %w", err)
```
Comment (danwt) [inline x/rollapp/keeper/keeper.go:84]: dont need failed to..

File: x/rollapp/keeper/keeper.go [hunk 6]
```diff
@@ -70,60 +73,33 @@ func (k Keeper) TriggerRollappGenesisEvent(ctx sdk.Context, rollapp types.Rollap
 	case rollapp.GenesisState.IsGenesisEvent:
 		return types.ErrGenesisEventAlreadyTriggered
 	}
+
+	// Register the denom metadata
```
Comment (danwt) [inline x/rollapp/keeper/keeper.go:77]: unnecessary comment

File: x/rollapp/keeper/keeper.go [hunk 7]
```diff
@@ -70,60 +73,33 @@ func (k Keeper) TriggerRollappGenesisEvent(ctx sdk.Context, rollapp types.Rollap
...
+	if err := k.registerDenomMetadata(ctx, rollapp); err != nil {
+		return fmt.Errorf("failed to register denom metadata: %w", err)
+	}
+
 	// Call the mint genesis tokens function
```
Comment (danwt) [inline x/rollapp/keeper/keeper.go:82]: comment could be deleted

File: x/rollapp/keeper/keeper.go [hunk 8]
```diff
@@ -150,11 +126,36 @@ func (k Keeper) RegisterDenomMetadata(ctx sdk.Context, rollapp types.Rollapp, ib
 			}
 			metadata.DenomUnits[j] = &newDu
 		}
+
 		// save the new token denom metadata
-		k.bankKeeper.SetDenomMetaData(ctx, metadata)
+		if err := k.denommetadataKeeper.CreateDenomMetadata(ctx, metadata); err != nil {
+			return fmt.Errorf("failed to create denom metadata: %w", err)
```
Comment (danwt) [inline x/rollapp/keeper/keeper.go:132]: dont need failed to

File: x/rollapp/keeper/keeper.go [hunk 9]
```diff
@@ -150,11 +126,36 @@ func (k Keeper) RegisterDenomMetadata(ctx sdk.Context, rollapp types.Rollapp, ib
...
+			return fmt.Errorf("failed to mint coins: %w", err)
+		}
+
+		accAddress, err := sdk.AccAddressFromBech32(acc.Address)
+		if err != nil {
+			return fmt.Errorf("failed to convert account address: %w", err)
+		}
+
+		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, accAddress, coinsToMint); err != nil {
+			return fmt.Errorf("failed to send coins to account: %w", err)
```
Comment (danwt) [inline x/rollapp/keeper/keeper.go:155]: dont need failed to

**Review by danwt — CHANGES_REQUESTED** [review]

**Comment by danwt** [comment]
CHANGELOG.md needed too

File: x/vfc/hooks/denom_metadata_hooks.go [hunk 10]
```diff
@@ -28,7 +29,9 @@ func (v VirtualFrontierBankContractRegistrationHook) AfterDenomMetadataCreation(
 	if strings.HasPrefix(strings.ToLower(newDenomMetadata.Base), ibcDenomPrefix) { // only deploy for IBC denom.
 		// Deploy the virtual frontier bank contract for the new IBC denom.
 		// Error, if any, no state transition will be made.
-		_ = v.evmKeeper.DeployVirtualFrontierBankContractForBankDenomMetadataRecord(ctx, newDenomMetadata.Base)
+		if err := v.evmKeeper.DeployVirtualFrontierBankContractForBankDenomMetadataRecord(ctx, newDenomMetadata.Base); err != nil {
+			return fmt.Errorf("failed to deploy virtual frontier bank contract for IBC denom %s: %w", newDenomMetadata.Base, err)
```
Comment (danwt) [inline x/vfc/hooks/denom_metadata_hooks.go:33]: dont need failed to 😄

**Review by danwt — APPROVED** [review]

File: CHANGELOG.md [hunk 11]
```diff
@@ -47,6 +47,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - (hygene) [#676](https://github.com/dymensionxyz/dymension/pull/676) lint tests
 - (ibc) [#678](https://github.com/dymensionxyz/dymension/pull/678) apply a pfm patch
 - (dependencies) [#677](https://github.com/dymensionxyz/dymension/pull/677) Bump cosmos ecosystem dependencies
+- (vfc) [#726](https://github.com/dymensionxyz/dymension/issues/726) Fix VFC hook not triggerred
```
Comment (mtsitrin) [inline CHANGELOG.md:50]: the main of this pr is the removal of `denometadata` as `ibc middleware` and change the registration to be `genesisEvent` based

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**